src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java

```diff
@@ -59,98 +50,85 @@ public GameStatus getStatus() {
-	 * all the non-author players
-	 */
-	private GameState assignCharacters() {
-		Function<String, Integer> randomAuthorOrderComparator = value ->
-				Double.valueOf(Math.random() * 1000).intValue();
+	Map<String, String> assignCharacters() {
+		Map<String, String> playerToCharacterCopy = new HashMap<>(this.playerCharacterMap);
 
-		final var authors =
-				this.suggestedCharacters.keySet()
-						.stream()
-						.sorted(Comparator.comparing(randomAuthorOrderComparator))
-						.collect(Collectors.toList());
+		Map<String, String> playerCharacterShuffled;
+		int countShuffledCharacters = 0;
 
-		authors.forEach(author -> {
-			final var character = this.getRandomCharacter()
-					.apply(this.suggestedCharacters.get(this.<String>cyclicNext().apply(authors, author)));
+		do {
+			playerCharacterShuffled = shuffledCharacters(playerToCharacterCopy);
+			countShuffledCharacters++;
+			if (countShuffledCharacters == 5) {
+				playerCharacterShuffled = moveCharacters(playerToCharacterCopy);
+			}
+		} while (!isTwoValueEquals(playerToCharacterCopy, playerCharacterShuffled) && countShuffledCharacters != 5);
 
-			character.markTaken();
+		return playerCharacterShuffled;
+	}
 
-			this.playerCharacterMap.put(author, character.getCharacter());
-		});
+	private Map<String, String> shuffledCharacters(Map<String, String> playerCharacter) {
+		List<String> key = new ArrayList<>(playerCharacter.keySet());
+		List<String> value = new ArrayList<>(playerCharacter.values());
 
-		final var authorsSet = new HashSet<>(authors);
+		Collections.shuffle(value);
 
-		final var nonTakenCharacters = this.suggestedCharacters.values()
-				.stream()
-				.flatMap(Collection::stream)
-				.filter(character -> !character.isTaken())
-				.collect(toList());
+		playerCharacter = IntStream.range(0, playerCharacter.size()).boxed()
+				.collect(Collectors.toMap(key::get, value::get));
 
-		this.players.keySet()
-				.stream()
-				.filter(player -> !authorsSet.contains(player))
-				.forEach(player -> {
-					final var character = this.getRandomCharacter().apply(nonTakenCharacters);
+		return playerCharacter;
+	}
 
-					character.markTaken();
+	private boolean isTwoValueEquals(Map<String, String> oldPlayerCharacter, Map<String, String> playerCharacterShuffled) {
+		int count = 0;
 
-					this.playerCharacterMap.put(player, character.getCharacter());
+		for (Map.Entry<String, String> entry : oldPlayerCharacter.entrySet()) {
+			String key = entry.getKey();
+			String val1 = entry.getValue();
+			String val2 = playerCharacterShuffled.get(key);
+			boolean isTwoValuesEqual = val1.equals(val2);
 
-					nonTakenCharacters.remove(character);
-				});
+			if (isTwoValuesEqual) {
+				count++;
+			}
+		}
 
-		return this;
+		return count == 0;
 	}
 
+	private static Map<String, String> moveCharacters(Map<String, String> playerCharacter) {
+		NavigableMap<String, String> tmpPlayerCharacter = new TreeMap<>(playerCharacter);
```

Comment:
It's great that you have implemented your own algorithm, but there is already created Java solution for that, https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#rotate(java.util.List,%20int), also I would like that shift would be random between 1 and `players.size()` and this number would be good to be as a parameter please look `moveCharacters(playerToCharacterCopy, new Random().nextInt(players.size() - 1) + 1)`, and this method also would be good to cover with the test and in test you will test shift by a specific number like `moveCharacters(playerToCharacterCopy, 2)`